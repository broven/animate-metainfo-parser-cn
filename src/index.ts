
import convert from 'xml-js';

export enum IResolution {
  rUHD = 'UHD',
  r1080p = '1080p',
  r720p = '720p',
  rUnknown = 'unknown'
}
export  const IResolutionPriority = {
  [IResolution.r1080p]: 10,
  [IResolution.r720p]: 5,
  [IResolution.rUHD]: 4,
  [IResolution.rUnknown]: 1
}
export class AnimateTitleMetaParser {
  private rawTitle: string;
  constructor(title: string) {
    this.rawTitle = title;
  }
  get resolution(): IResolution {
    const t = this.rawTitle;
    switch(true) {
      case t.indexOf('1080p') !== -1:
      case t.indexOf('1080P') !== -1:
      case t.indexOf('1920X1080') !== -1:
        return IResolution.r1080p;
      case t.indexOf('720p') !== -1:
      case t.indexOf('1280X720') !== -1:
        return IResolution.r720p;
      default:
        return IResolution.rUnknown;
    }
  }
  get isSeasonPack(): boolean {
    const title = this.rawTitle
    switch(true) {
        case title.indexOf('【合集】') !== -1:
            return true;
        default:
            return false;
    }
  }
  get isTraditionalChinese(): boolean {
    const text = this.rawTitle;
    switch(true) {
        case text.indexOf('BIG5') !== -1:
        case text.indexOf('CHT') !== -1:
            return true;
        default:
            return false;
    }
  }
  get season(): number {
    return -1;
  }
  get episode(): number {
    const splitterArr = [
      /】/, /【/, /\[/, /\]/, /\s/
    ];
    const regArr = [
      /~~~(\d{1,})~~~/
    ];
    let title = this.rawTitle;
    for (const splitter of splitterArr) {
      const _reg = new RegExp(splitter, 'g');
      title = title.replace(_reg, '~~~');
    }
    for (const reg of regArr) {
      const match = reg.exec(title);
      if (match) {
        return parseInt(match[1]);
      }
    }
    return -1;
  }
  get videoId(): string {
    return `S${this.season}-E${this.episode}`
  }
}

export const animateFilter = (content: string) => {
  const parsedFeed:any = convert.xml2js(content, { compact: true});
  parsedFeed.rss.channel.item = parsedFeed.rss.channel.item.map((item: any) => {
    const title = item.title._text;
    const metaParser = new AnimateTitleMetaParser(title);
    if (metaParser.isSeasonPack) return null;
    if (metaParser.isTraditionalChinese) return null;
   return item;
  })
  .filter((i: any) => i !== null)

  parsedFeed.rss.channel.item = parsedFeed.rss.channel.item
  .filter((item: any) => {
    // 过滤相同视频最高分辨率
    const itemMeta = new AnimateTitleMetaParser(item.title._text);
    const isExistHigherResolution = (parsedFeed.rss.channel.item as any[]).findIndex(i => {
      const title = i.title._text;
      const metaParser = new AnimateTitleMetaParser(title);
      if (metaParser.videoId === itemMeta.videoId) {
        if (IResolutionPriority[metaParser.resolution] > IResolutionPriority[itemMeta.resolution]) {
          return true;
        }
      }
      return false;
    });
    if (isExistHigherResolution !== -1) return null;
    return item;
  })
  .filter((i: any) => i !== null)

  return convert.js2xml(parsedFeed, {compact: true, spaces: 4});
}