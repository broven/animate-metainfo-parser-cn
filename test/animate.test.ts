// Copyright (C) 2021 metajs
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import { animateFilter, AnimateTitleMetaParser, IResolution } from '../src';
import fs from 'fs';
// https://mikanani.me/
describe('幻樱字幕组', () => {
    // test('幻樱字幕组', () => {
    //     const content = fs.readFileSync('test/data/幻樱字幕组.rss');
    //     const result = animateFilter(content.toString());
    //     expect(result).toMatchSnapshot();
    // });
    test('合集', () => {
        const content = fs.readFileSync('test/data/collection.rss');
        const result = animateFilter(content.toString());
        expect(result).toMatchSnapshot();
    });
    test('语言过滤', () => {
        const content = fs.readFileSync('test/data/language.rss');
        const result = animateFilter(content.toString());
        expect(result).toMatchSnapshot();
    });
    test('单集多分辨率', () => {
        const content = fs.readFileSync('test/data/uni.rss');
        const result = animateFilter(content.toString());
        expect(result).toMatchSnapshot();
    });

})
describe.each([
    ['【动漫国字幕组】★07月新番[碧水白沙 / 白沙的水族馆][11-14][1080P][繁体][MP4]', IResolution.r1080p, false, false, -1, -1], // 这种的集数简直毒瘤
    ['[喵萌奶茶屋&LoliHouse] 平家物语 / Heike Monogatari - 03 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]', IResolution.r1080p, false, false, -1, 3],
    ['【喵萌Production】★07月新番★[Love Live! Superstar!! / ラブライブ！スーパースター!!][11][1080p][繁日双语][招募翻译]', IResolution.r1080p, false, false, -1, 11],
])('AnimateTitleMetaParser %s', (title, resolution, isSeasonPack, isChinese, season, ep) => {
    const metaParser = new AnimateTitleMetaParser(title);
    test('分辨率', () => {
        expect(metaParser.resolution).toBe(resolution);
    })
    test('是否是合集', () => {
        expect(metaParser.isSeasonPack).toBe(isSeasonPack);
    })
    test('中文识别', () => {
          expect(metaParser.isTraditionalChinese).toEqual(isChinese)
    })
    test('season', () => {
        expect(metaParser.season).toBe(season);
    });
    test('episode', () => {
        expect(metaParser.episode).toBe(ep);
    })

})