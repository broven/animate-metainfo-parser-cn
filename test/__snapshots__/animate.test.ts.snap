// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`幻樱字幕组 单集多分辨率 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <item>
            <title>【幻樱字幕组】【10月新番】【养猫养狗每天很开心 Inu to Neko Docchi mo Katteru to Mainichi Tanoshii】【01】【GB_MP4】【1920X1080】</title>
        </item>
    </channel>
</rss>"
`;

exports[`幻樱字幕组 合集 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>Mikan Project - 养猫养狗每天很开心</title>
        <link>http://mikanani.me/RSS/Bangumi?bangumiId=2300&amp;subgroupid=21</link>
        <description>Mikan Project - 养猫养狗每天很开心</description>
    </channel>
</rss>"
`;

exports[`幻樱字幕组 语言过滤 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>Mikan Project - 养猫养狗每天很开心</title>
        <link>http://mikanani.me/RSS/Bangumi?bangumiId=2300&amp;subgroupid=21</link>
        <description>Mikan Project - 养猫养狗每天很开心</description>
        <item>
            <guid isPermaLink=\\"false\\">【幻樱字幕组】【10月新番】【养猫养狗每天很开心 Inu to Neko Docchi mo Katteru to Mainichi Tanoshii】【01】【GB_MP4】【1280X720】</guid>
            <link>https://mikanani.me/Home/Episode/746872bfc27c1d89123de4b0351d99e7ba4eb12a</link>
            <title>【幻樱字幕组】【10月新番】【养猫养狗每天很开心 Inu to Neko Docchi mo Katteru to Mainichi Tanoshii】【01】【GB_MP4】【1280X720】</title>
            <description>【幻樱字幕组】【10月新番】【养猫养狗每天很开心 Inu to Neko Docchi mo Katteru to Mainichi Tanoshii】【01】【GB_MP4】【1280X720】[14MB]</description>
            <torrent xmlns=\\"https://mikanani.me/0.1/\\">
                <link>https://mikanani.me/Home/Episode/746872bfc27c1d89123de4b0351d99e7ba4eb12a</link>
                <contentLength>14680064</contentLength>
                <pubDate>2020-10-03T19:13:00</pubDate>
            </torrent>
            <enclosure type=\\"application/x-bittorrent\\" length=\\"14680064\\" url=\\"https://mikanani.me/Download/20201003/746872bfc27c1d89123de4b0351d99e7ba4eb12a.torrent\\"/>
        </item>
    </channel>
</rss>"
`;
